File: 2022/day08/javascript/part1.mjs
import { join } from "node:path";
import { readFile } from "node:fs/promises";

const filePath = new URL(join(import.meta.url, "..", "..", "input.txt"));

const file = await readFile(filePath, { encoding: "utf-8" });
const data = file.split("\n");

let visible = data[0].length * 2 + data.length * 2 - 4;
const treeMap = data.map((row) => row.split("").map(Number));

for (let i = 1; i < data.length - 1; i++) {
  for (let j = 1; j < data[i].length - 1; j++) {
    const tree = treeMap[i][j];
    const leftRow = [...treeMap[i]].splice(0, j);
    const rightRow = [...treeMap[i]].splice(j + 1, data[i].length);

    if (Math.max(...leftRow) < tree || Math.max(...rightRow) < tree)
      visible += 1;
    else {
      const topList = [];
      const bottomList = [];

      for (let k = 0; k < i; k++) {
        topList.push(treeMap[k][j]);
      }

      for (let k = i + 1; k < treeMap.length; k++) {
        bottomList.push(treeMap[k][j]);
      }

      if (Math.max(...topList) < tree || Math.max(...bottomList) < tree)
        visible += 1;
    }
  }
}

console.log(visible);

File: 2022/day08/python/part1.py
from pathlib import Path
file_path = Path(__file__, '..', '..', 'input.txt').resolve()


with open(file_path) as f:
    data = f.read().splitlines()

    visible = len(data[0]) * 2 + len(data) * 2 - 4
    tree_map = [list(map(int, x)) for x in data]

    for i in range(1, len(tree_map) - 1):
        for j in range(1, len(tree_map[i]) - 1):
            tree = tree_map[i][j]
            left_row = tree_map[i][:j]
            right_row = tree_map[i][j + 1:]

            if max(left_row) < tree or max(right_row) < tree:
                visible += 1

            else:
                top_list = []
                bottom_list = []

                for k in range(0, i):
                    top_list.append(tree_map[k][j])

                for k in range(i + 1, len(tree_map)):
                    bottom_list.append(tree_map[k][j])

                if max(top_list) < tree or max(bottom_list) < tree:
                    visible += 1

    print(visible)
